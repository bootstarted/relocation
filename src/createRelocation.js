// @flow
import * as React from 'react';
import cuid from 'cuid';

type Item = {id: string, child: React.Node, meta: *};
type Elements = {[string]: {children: React.Node, meta: *}};
type ItemMapper = ($ReadOnlyArray<Item>) => React.Node;

const flattenElements = (elements: Elements): $ReadOnlyArray<Item> => {
  const results = [];
  Object.keys(elements).forEach((id) => {
    const {children, meta} = elements[id];
    React.Children.forEach(children, (child) => {
      results.push({child, meta, id});
    });
  });
  return results;
};

class RelocationDelegate extends React.PureComponent<{
  children: React.Node,
  removeChild: *,
  addChild: *,
  meta?: *,
}> {
  id = cuid();

  componentWillUnmount() {
    this.props.removeChild(this.id);
  }

  componentDidUpdate() {
    this.props.addChild(this.id, this.props.children, this.props.meta);
  }

  componentDidMount() {
    this.props.addChild(this.id, this.props.children, this.props.meta);
  }

  render() {
    return null;
  }
}

const createRelocation = () => {
  const RelocationInjector = React.createContext({
    addChild: () => {},
    removeChild: () => {},
  });

  const RelocationChildren = React.createContext({});

  class Collector {
    elements: Elements = {};
    mapItems: ItemMapper;
    constructor(mapItems: ItemMapper) {
      this.mapItems = mapItems;
    }
    onAdd(id: string, children: React.Node, meta: *) {
      this.elements = {
        ...this.elements,
        [id]: {children, meta},
      };
    }
    getElements() {
      return this.mapItems(flattenElements(this.elements));
    }
  }

  class Provider extends React.Component<
    {
      children: React.Node,
      collector?: Collector,
    },
    {
      elements: Elements,
    },
  > {
    state = {
      elements: {},
    };
    elements = {};

    addChild = (id: string, children: *, meta: *) => {
      this.elements = {
        ...this.elements,
        [id]: {children, meta},
      };
      this.setState({elements: this.elements});
      if (this.props.collector) {
        this.props.collector.onAdd(id, children, meta);
      }
    };

    removeChild = (id: string) => {
      if (id in this.elements) {
        const elements = {...this.elements};
        delete elements[id];
        this.elements = elements;
        this.setState({elements: this.elements});
      }
    };

    headInjectorContext = {
      addChild: this.addChild,
      removeChild: this.removeChild,
    };

    render() {
      return (
        <RelocationInjector.Provider value={this.headInjectorContext}>
          <RelocationChildren.Provider value={this.state.elements}>
            {this.props.children}
          </RelocationChildren.Provider>
        </RelocationInjector.Provider>
      );
    }
  }

  const Component = ({children, meta}: {children: React.Node, meta?: *}) => (
    <RelocationInjector.Consumer>
      {(context) => (
        <RelocationDelegate
          meta={meta}
          children={children}
          addChild={context.addChild}
          removeChild={context.removeChild}
        />
      )}
    </RelocationInjector.Consumer>
  );

  const Consumer = ({children}: {children: ItemMapper}) => (
    <RelocationChildren.Consumer>
      {(context) => children(flattenElements(context))}
    </RelocationChildren.Consumer>
  );

  return {Consumer, Collector, Provider, Component};
};

export default createRelocation;
