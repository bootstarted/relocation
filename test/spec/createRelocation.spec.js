import * as React from 'react';
import renderer from 'react-test-renderer';

import createRelocation from '../../src/createRelocation';

const Relocation = createRelocation();

class Tester extends React.Component {
  static defaultProps = {
    value: 'Hello',
  };
  state = {enabled: false};
  on() {
    this.setState({enabled: true});
  }
  off() {
    this.setState({enabled: false});
  }
  render() {
    const {enabled} = this.state;
    return (
      <Relocation.Provider collector={this.props.collector}>
        <Relocation.Consumer>
          {(elements) => {
            if (elements[0]) {
              return elements[0].child;
            }
            return null;
          }}
        </Relocation.Consumer>

        {enabled && (
          <Relocation.Component>
            <div>{this.props.value}</div>
          </Relocation.Component>
        )}
      </Relocation.Provider>
    );
  }
}

describe('createRelocation', () => {
  it('should portal elements in and out of the tree', () => {
    const component = renderer.create(<Tester />);
    expect(component.toJSON()).toMatchSnapshot();
    component.root.instance.on();
    expect(component.toJSON()).toMatchSnapshot();
  });
  it('should work with collector', () => {
    const collector = new Relocation.Collector((x) => x[0].child);
    const component = renderer.create(<Tester collector={collector} />);
    component.root.instance.on();
    const component2 = renderer.create(collector.getElements());
    expect(component2.toJSON()).toMatchSnapshot();
  });
  it('should unmount things', () => {
    const component = renderer.create(<Tester />);
    component.root.instance.on();
    component.unmount();
  });
  it('should handle component prop updates', () => {
    const component = renderer.create(<Tester key="5" />);
    component.root.instance.on();
    component.update(<Tester key="5" value="spagoots" />);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
