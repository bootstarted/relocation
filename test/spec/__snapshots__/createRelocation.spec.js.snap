// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createRelocation should handle component prop updates 1`] = `
<div>
  spagoots
</div>
`;

exports[`createRelocation should portal elements in and out of the tree 1`] = `null`;

exports[`createRelocation should portal elements in and out of the tree 2`] = `
<div>
  Hello
</div>
`;

exports[`createRelocation should work with collector 1`] = `
<div>
  Hello
</div>
`;
